### PR TITLE
[CESQL] Add ANTLR4 grammar

### PR DESCRIPTION
**Pull request by slinkydeveloper**
This PR includes the lexer and grammar specs to implement CESQL in the ANTLR4 format. 

It also includes some small fixes/typos I made to the spec found out while working on the grammar

### CODE REVIEW

**Comment by slinkydeveloper** [comment]
cc @lionelvillard

File: CESQLParser.g4 [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+notOperator
+    : NOT
+    ;
+
+unaryLogicOperator
+    : notOperator
+    ;
+
+binaryComparisonOperator
+    : (EQUAL | EXCLAMATION EQUAL | LESS GREATER | GREATER EQUAL | LESS EQUAL | LESS | GREATER)
```
Comment (lionelvillard) [inline CESQLParser.g4:70]: Usually comparison operators don't allow spaces between characters, e.g `<     =` vs `<=`

File: CESQLParser.g4 [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    ;
+
+functionParameterList
+    : ( functionParameter ( COMMA functionParameter )* )?
+    ;
+
+// Sets
+
+setExpression
+    : expression ( COMMA expression )* // Non-empty set is not allowed
```
Comment (lionelvillard) [inline CESQLParser.g4:102]: maybe: `empty sets are not allowed`

File: CESQLParser.g4 [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    : unaryLogicOperator expression #unaryLogicExpression
+    | unaryNumericOperator expression # unaryNumericExpression
+    ;
+
+expression
+    : functionInvocation #functionInvocationExpression
+    | unaryOperation #unaryExpression
+    // LIKE, EXISTS and IN takes precedence over all the other operators
+    | expression notOperator? likeOperator stringLiteral #likeExpression
+    | existsOperator identifier #existsExpression
```
Comment (lionelvillard) [inline CESQLParser.g4:23]: what about inlining the `xxOperator` rules? example: `expression NOT? LIKE stringLiteral`. I find this easier to read and it's consistent with what you did below.

**Comment by slinkydeveloper** [comment]
FYI here is my first draft impl: https://github.com/cloudevents/sdk-java/pull/363

**Comment by duglin** [comment]
@grant can you explain the lint commit error messages?

**Comment by grant** [comment]
> @grant can you explain the lint commit error messages?

Yes. The goal of linting the commits is to get the commits in conventional commit (or at least the last one).

A commit message like `fix(CESQL): function invocation rule` should make the test pass.

**Comment by slinkydeveloper** [comment]
@grant is it something we really need? Couldn't that be done while squashing?

**Comment by grant** [comment]
> @grant is it something we really need? Couldn't that be done while squashing?

Yes, you can so it while squashing, or by updating the latest commit.

**Comment by duglin** [comment]
Approved on the 4/1 call but let's wait until we make the linter happy before we merge

**Comment by slinkydeveloper** [comment]
@duglin you can make the linter happy while squashing :)

**Pull request merged**